### ISSUE TITLE
Aggregation tools: Reducer

### ISSUE THREAD

**Issue by cavities** [python-bonobo/bonobo#245]
[need] i think bonobo need  aggregation tools

**Comment by hartym** [comment]
yes :D

**Comment by Autoplectic** [comment]
What exactly is meant by aggregation tools?  I would be interest in working on a "join" node, that takes all the output of the prior node in the chain as a single input. If this is specifically outside the scope of bonobo, I understand that as well.

From looking around, it looks like to optimal location for adding this would be in NodeExecutionContext._get() or similar.

**Comment by hartym** [comment]
I think this would need a RFC/design doc before. Feel free to go ahead and start one, good place to discuss such a feature: https://github.com/python-bonobo/bonobo/wiki

There are different aggregations we can think off:

- reduce-style aggregation (initial value + reduce function?)
- database like aggregation (group-by + agg func)
- other ?

“join” nodes are tricky, I'm not sure I understand what you're describing.

If we talk about joining different "streams" (from input (A,B), take (a1,b2)), and that's not as easy as the concept sound because there is no sync between different queues (how do we know we'll get the same amount of data each side ? How do we avoid one node accidentally skipping a row because error?).

When you say all outputs of prior as a single input, how doses this differ from an "infinite width" fixed window? Or maybe I did not get what you said?

**Comment by levic** [comment]
> reduce-style aggregation

This is what I was looking for. It took me a few hours to work out how you could go about doing this.

- For reference for anyone else looking at this ticket, here is something that achieves this; it tries to be a direct counterpart for `functools.reduce`. Code is adapted from `FixedWindow` -- there well may be a simpler way to implement this.

```
# This can't be used directly, you want to override `function` (and maybe `initializer`)
# `function` and `initializer` have the same meaning as in `functools.reduce`
class Reducer(bonobo.config.Configurable):
    initializer = None

    def function(self):
        raise NotImplementedError('function must be specified to use Reducer')

    def reduce(self, values, initializer):
        return functools.reduce(self.function, values, initializer)

    @bonobo.config.ContextProcessor
    def buffer(self, context):
        values = yield []
        initializer = self.initializer() if callable(self.initializer) else self.initializer
        reduced_value = self.reduce(values, initializer)
        context.send(reduced_value)

    @bonobo.config.use_raw_input
    def __call__(self, values, bag):
        values.append(bag)

# This can be used directly by passing `function` and `initializer` as constructor options
class ConfigurableReducer(Reducer):
    function = bonobo.config.options.Option(None, positional=True)
    initializer = bonobo.config.options.Option(None, positional=True, required=False)
```

**Comment by hartym** [comment]
Thanks for the example!

This looks a very nice enhancement that could benefit *a lot* to the bonobo core (although I did not try it / test it yet, but conceptually, I think it is right).

One defect that happens to exist in this implementation though is that it needs to "buffer" the whole stream content before starting to reduce. I hope not to overlook it by saying that it's not necessary, as `reduce(f, a+b)` should be equivalent to `reduce(f, a) + reduce(f, b)`, and thus it should be possible to live-reduce the values (although the result can only be sent at the end).

I also like the fact you separate `Reducer` and `ConfigurableReducer`, this is a practice that could be extended to a lot of the bonobo core, although I don't understand why `Reducer` extends `Configurable` (and why it's not multiple inheritance in `ConfigurableReducer`). Is that because of a defect in bonobo that forbids you to do so?

**Comment by levic** [comment]
> it needs to "buffer" the whole stream content before starting to reduce

Good point, I only had a small amount of data that I was reducing so I didn't look into doing this efficiently -- it certainly doesn't need to buffer everything before reducing.

>  I don't understand why Reducer extends Configurable

Ah. It doesn't need to, I tried to do everything in one class but ran into #266 so split it out into two classes - I wanted to be able to define `initializer` directly in a new Reducer class.

**Comment by hartym** [comment]
I used and changed a bit your code in a transformation, works well for me.

```
class Reducer(Configurable):
    function = Method()
    initializer = Option(required=False)

    def reduce(self, values, initializer):
        return functools.reduce(self.function, values, initializer)

    @ContextProcessor
    def buffer(self, context):
        values = yield []
        initializer = self.initializer() if callable(self.initializer) else self.initializer
        reduced_value = self.reduce(values, initializer)
        context.send(reduced_value)

    @use_raw_input
    def __call__(self, values, bag):
        values.append(bag)
```

The only main difference is that the reducer function now takes "self" as the first argument, which is not what we may want. This is a defect in `Method` and I'm looking at a way to fix that. But with this code, I can do:

```
def some_reducer(self, left, right):
    return ...

graph.add_chain(..., Reducer(some_reducer, initializer={...}), ...)
```

I'm looking forward to integrating this in next bonobo version, if you want to do a pull request yourself to include you in the contributors, go ahead, otherwise I'll do it and reference you in the commit.

**Comment by hartym** [comment]
Integrated in https://github.com/python-bonobo/bonobo/blob/develop/bonobo/nodes/aggregation.py
